File: services/tracking/src/features/tracking/commands/e2e_cleanup.py
"""`DELETE /v1/trackings/e2e-cleanup` command handler (JE-111).

The E2E suite's teardown: soft-deletes the trackings the harness created, so a
rerun starts from a clean slate. The peer of Orders' `Endpoints/E2eEndpoints.cs`
and Users' `http/e2e-cleanup.ts`.

Transport-free like every other command here — it takes a session and returns a
count, so the router is a thin translation.

## Deleted by TAG, not by caller

It was scoped by the caller's `cognito_sub` first, and that could not work: the
teardown runs ONCE at the end of a suite, globally, with no user session and
therefore no `x-user-id` header — so the endpoint's only real caller would have
been refused `401` by the very dependency that made the scoping possible.

Tagging at creation moves the decision to the one moment an identity IS present
(`shared/http/e2e_source.py`), and leaves this a single unscoped predicate: delete
what the harness made. Users' cleanup already worked this way, on the same literal
tag value.

Losing the caller scope loses nothing, because the safety it provided moved rather
than disappeared:

* a row is tagged only when the request said so **and** `E2E_TESTING_ENABLED` is
  on, so an untrusted client cannot enlist its own rows;
* this route is mounted only under that same flag;
* every row a real user ever created is untagged, and therefore untouchable here.

## Soft-delete, never a hard delete

Deleting stamps `deleted_at`/`deleted_by` ([[soft-delete]]). A test-support
endpoint is exactly where a "just DELETE the rows, it is only test data" shortcut
looks reasonable, and it is the one place it must not be taken: this route runs
against the SAME schema and the same database user as production, which is
granted no `DELETE` privilege at all. The rows stay, invisible to every read.

## The actor is the harness, not the caller

`AuditActor.E2E_CLEANUP` rather than any identity off the request — there is no
longer an identity on the request at all, but the reasoning predates that: a row
removed by the test harness must stay distinguishable from one removed by a real
flow, which is precisely the distinction `deleted_by` exists to record. Both
Orders and Users make the same call.
"""

from __future__ import annotations

from sqlalchemy.orm import Session

from src.features.tracking.domain.models import E2E_SOURCE_TAG
from src.features.tracking.domain.repository import TrackingRepository
from src.shared.audit.audit_actor import AuditActor


def e2e_cleanup(session: Session, *, tag: str = E2E_SOURCE_TAG) -> int:
    """Soft-delete every live tracking carrying `tag`; return how many were stamped.

    The tag is a parameter with a default rather than a constant read inside, so
    the behaviour is stated at the boundary and a test can pin "only this tag" by
    seeding another one — but no caller has to know the literal.

    Does not commit — the caller's `write_session` owns the transaction boundary,
    exactly as `create_tracking` and `update_tracking_status` do.

    Idempotent: a second call stamps nothing (the rows are no longer live) and
    returns `0`, which is a success, not an error. The count is returned to the
    client so a teardown that quietly matched nothing is visible in the harness's
    own output rather than only in this service's logs.
    """
    return TrackingRepository(session).soft_delete_by_tag(
        tag, actor=AuditActor.E2E_CLEANUP
    )

File: services/tracking-go/internal/app/e2e_cleanup.go
package app

import (
	"context"
	"time"

	"github.com/jemartinez/3mrai/services/tracking-go/internal/domain"
	"github.com/jemartinez/3mrai/services/tracking-go/internal/domain/audit"
)

// TagSoftDeleter is the E2E teardown's port, unscoped by identity because the
// teardown runs once globally with no user session.
//
// CONTRACT: The scoping lives elsewhere and BOTH halves are required — a row is
// tagged only under x-e2e-source AND E2E_TESTING_ENABLED, and this route does
// not exist without that flag. A real user's rows are untagged.
// See [[testing]]
type TagSoftDeleter interface {
	SoftDeleteByTag(ctx context.Context, tag string, actor audit.Actor, now time.Time) (int64, error)
	// SoftDeleteByTags requires BOTH tags. An empty secondTag means "no scope"
	// and must behave exactly like SoftDeleteByTag.
	SoftDeleteByTags(ctx context.Context, tag, secondTag string, actor audit.Actor, now time.Time) (int64, error)
}

// E2ECleanup is the E2E suite's teardown.
type E2ECleanup struct {
	deleter TagSoftDeleter
	clock   func() time.Time
}

// NewE2ECleanup wires the use case. The nil-clock fallback mints the same
// whole-second UTC timestamp every write path in this service uses.
func NewE2ECleanup(deleter TagSoftDeleter, clock func() time.Time) *E2ECleanup {
	if clock == nil {
		clock = func() time.Time { return time.Now().UTC().Truncate(time.Second) }
	}
	return &E2ECleanup{deleter: deleter, clock: clock}
}

// Execute soft-deletes every live tracking carrying the E2E Source tag. It is
// idempotent, and the count reaches the client so a teardown matching nothing
// shows in the harness's output.
//
// CONTRACT: The tag comes from the domain constant, never a literal here — a
// near-miss cleans up nothing while looking correct.
func (uc *E2ECleanup) Execute(ctx context.Context) (int64, error) {
	return uc.ExecuteScoped(ctx, "")
}

// ExecuteScoped soft-deletes the E2E rows of ONE run. An empty runTag deletes
// every E2E row, which is Execute's behaviour.
//
// CONTRACT: Scope a parallel run's teardown. The unscoped sweep deletes every
// E2E-tagged tracking on the machine, so one teardown lands inside another run's
// live progression, whose next tick reads tracking_not_found and aborts —
// the remaining statuses never publish, which reads as a lost message.
// See [[2026-08-30-a-global-teardown-cannot-be-scoped]]
func (uc *E2ECleanup) ExecuteScoped(ctx context.Context, runTag string) (int64, error) {
	return uc.deleter.SoftDeleteByTags(ctx, domain.E2ESourceTag, runTag, audit.E2ECleanup, uc.clock())
}
